staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go

```diff
@@ -135,6 +137,130 @@ func (e *watchEmbeddedEncoder) embeddedIdentifier() runtime.Identifier {
+		embeddedEncoder: embeddedEncoder,
+		encoder:         encoder,
+		framer:          framer,
+		buffer:          runtime.NewSpliceBuffer(),
+		internalEvent:   &metav1.InternalEvent{},
+		outEvent:        &metav1.WatchEvent{},
+		eventBuffer:     runtime.NewSpliceBuffer(),
+	}
+}
+
+// Encode encodes a given watch event.
+// NOTE: if events object is implementing the CacheableObject interface,
+//
+//	the serialized version is cached in that object [not the event itself].
+func (e *watchEncoder) Encode(event watch.Event) error {
+	encodeFunc := func(obj runtime.Object, w io.Writer) error {
+		return e.doEncode(obj, event, w)
+	}
+	if co, ok := event.Object.(runtime.CacheableObject); ok {
+		return co.CacheEncode(e.identifier(event.Type), encodeFunc, e.framer)
+	}
+	return encodeFunc(event.Object, e.framer)
+}
+
+func (e *watchEncoder) doEncode(obj runtime.Object, event watch.Event, w io.Writer) error {
+	defer e.buffer.Reset()
+
+	if err := e.embeddedEncoder.Encode(obj, e.buffer); err != nil {
+		return fmt.Errorf("unable to encode watch object %T: %v", obj, err)
+	}
+
+	// ContentType is not required here because we are defaulting to the serializer type.
+	e.unknown.Raw = e.buffer.Bytes()
+	event.Object = &e.unknown
+	defer func() {
+		// Restore the original event to not modify the object as part of Encode.
+		event.Object = obj
+	}()
+	metrics.WatchEventsSizes.WithContext(e.ctx).WithLabelValues(e.kind.Group, e.kind.Version, e.kind.Kind).Observe(float64(len(e.unknown.Raw)))
+
+	*e.outEvent = metav1.WatchEvent{}
+
+	// create the external type directly and encode it.  Clients will only recognize the serialization we provide.
+	// The internal event is being reused, not reallocated so its just a few extra assignments to do it this way
+	// and we get the benefit of using conversion functions which already have to stay in sync
+	*e.internalEvent = metav1.InternalEvent(event)
+	if err := metav1.Convert_v1_InternalEvent_To_v1_WatchEvent(e.internalEvent, e.outEvent, nil); err != nil {
+		return fmt.Errorf("unable to convert watch object: %v", err)
+	}
+
+	defer e.eventBuffer.Reset()
+	if err := e.encoder.Encode(e.outEvent, e.eventBuffer); err != nil {
+		return fmt.Errorf("unable to encode watch object %T: %v (%#v)", e.outEvent, err, e)
+	}
+
+	_, err := w.Write(e.eventBuffer.Bytes())
+	return err
+}
+
+type watchEncoderIdentifier struct {
+	Name            string `json:"name,omitempty"`
+	EmbeddedEncoder string `json:"embeddedEncoder,omitempty"`
+	Encoder         string `json:"encoder,omitempty"`
+	EventType       string `json:"eventType,omitempty"`
+}
+
+func (e *watchEncoder) identifier(eventType watch.EventType) runtime.Identifier {
+	// We need to take into account that in embeddedEncoder includes table
+	// transformer, then its identifier is dynamic. As a result, whenever
+	// the identifier of embeddedEncoder changes, we need to invalidate the
+	// whole identifiers cache.
+	// TODO(wojtek-t): Can we optimize it somehow?
```

Comment:
is this PR still more efficient, even before this is optimized?

Reply:
It is - please take a look at the PR description for benchmarks numbers [I just updated them with the changes from this review comment round]